### DISCUSSION TITLE
Module 'pluggy' not found

### DISCUSSION THREAD

**Discussion by farhills** [R0Wi-DEV/workflow_ocr#181] [General]
This is more likely a support request, but I don't know a better place to ask.

I'm on NC 25.0.3; Workflow_OCR 1.25.3

Docker environment, alpine linux (Linuxserver container : latest)

I have ocrmypdf correctly installed in the container. Confirmed by running `ocrmypdf -v --redo-ocr test.pdf output.pdf` from a user folder, executed within the container.  A new copy of the file is correctly generated with the OCR layer added.

When running via cron, there is an error with the 'pluggy' module not found. I'm guessing this has something to do with the docker environment, and files not being located where expected.  But I've reached the end of my capability to investigate.

Excerpt:
`[...]File \"/usr/lib/python3.9/site-packages/ocrmypdf/__init__.py\", line 8, in <module>\n    from pluggy import HookimplMarker as _HookimplMarker\nModuleNotFoundError: No module named 'pluggy',[...]`

Full Log:
```
{"reqId":"zo00fPcm1UTyaxeOOfXe","level":0,"time":"2023-01-30T22:02:30+00:00","remoteAddr":"","user":"--","app":"cron","method":"","url":"--","message":"CLI cron call has selected job with ID 12930","userAgent":"--","version":"25.0.3.2","data":{"app":"cron"}}
{"reqId":"zo00fPcm1UTyaxeOOfXe","level":0,"time":"2023-01-30T22:02:30+00:00","remoteAddr":"","user":"--","app":"cron","method":"","url":"--","message":"Run OCA\\WorkflowOcr\\BackgroundJobs\\ProcessFileJob job with ID 12930","userAgent":"--","version":"25.0.3.2","data":{"app":"cron"}}
{"reqId":"zo00fPcm1UTyaxeOOfXe","level":0,"time":"2023-01-30T22:02:30+00:00","remoteAddr":"","user":"--","app":"workflow_ocr","method":"","url":"--","message":"STARTED -- Run OCA\\WorkflowOcr\\BackgroundJobs\\ProcessFileJob job. Argument: {\"filePath\":\"/[user redacted]/files/Misc-JD/OCR-Testing/iiiii - Copy - Copy - Copy.pdf\",\"settings\":\"{\\\"languages\\\":[\\\"eng\\\"],\\\"ocrMode\\\":1,\\\"removeBackground\\\":false}\"}.","userAgent":"--","version":"25.0.3.2","data":{"app":"workflow_ocr"}}
{"reqId":"D8cLtcf6aEPvaG4LYxOf","level":0,"time":"2023-01-30T22:02:30+00:00","remoteAddr":"[user local IP]","user":"[user redacted]","app":"metadata","method":"GET","url":"/apps/logreader/poll?lastReqId=bJn6LLHkWJKCPFrSAxZz","message":"/appinfo/app.php is deprecated, use \\OCP\\AppFramework\\Bootstrap\\IBootstrap on the application class instead.","userAgent":"Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/109.0.0.0 Safari/537.36","version":"25.0.3.2","data":{"app":"metadata"}}
{"reqId":"zo00fPcm1UTyaxeOOfXe","level":0,"time":"2023-01-30T22:02:30+00:00","remoteAddr":"","user":"[user redacted]","app":"workflow_ocr","method":"","url":"--","message":"Running command: ocrmypdf -q --redo-ocr -l eng --sidecar /tmp/oc_tmp_okdJpH-.sidecar - - | cat","userAgent":"--","version":"25.0.3.2","data":{"app":"workflow_ocr"}}
{"reqId":"zo00fPcm1UTyaxeOOfXe","level":2,"time":"2023-01-30T22:02:30+00:00","remoteAddr":"","user":"[user redacted]","app":"workflow_ocr","method":"","url":"--","message":"OCRmyPDF succeeded with warning(s): Traceback (most recent call last):\n  File \"/usr/bin/ocrmypdf\", line 33, in <module>\n    sys.exit(load_entry_point('ocrmypdf==12.7.1', 'console_scripts', 'ocrmypdf')())\n  File \"/usr/bin/ocrmypdf\", line 25, in importlib_load_entry_point\n    return next(matches).load()\n  File \"/usr/lib/python3.9/importlib/metadata.py\", line 86, in load\n    module = import_module(match.group('module'))\n  File \"/usr/lib/python3.9/importlib/__init__.py\", line 127, in import_module\n    return _bootstrap._gcd_import(name[level:], package, level)\n  File \"<frozen importlib._bootstrap>\", line 1030, in _gcd_import\n  File \"<frozen importlib._bootstrap>\", line 1007, in _find_and_load\n  File \"<frozen importlib._bootstrap>\", line 972, in _find_and_load_unlocked\n  File \"<frozen importlib._bootstrap>\", line 228, in _call_with_frames_removed\n  File \"<frozen importlib._bootstrap>\", line 1030, in _gcd_import\n  File \"<frozen importlib._bootstrap>\", line 1007, in _find_and_load\n  File \"<frozen importlib._bootstrap>\", line 986, in _find_and_load_unlocked\n  File \"<frozen importlib._bootstrap>\", line 680, in _load_unlocked\n  File \"<frozen importlib._bootstrap_external>\", line 850, in exec_module\n  File \"<frozen importlib._bootstrap>\", line 228, in _call_with_frames_removed\n  File \"/usr/lib/python3.9/site-packages/ocrmypdf/__init__.py\", line 8, in <module>\n    from pluggy import HookimplMarker as _HookimplMarker\nModuleNotFoundError: No module named 'pluggy', ","userAgent":"--","version":"25.0.3.2","data":{"app":"workflow_ocr"}}
{"reqId":"zo00fPcm1UTyaxeOOfXe","level":3,"time":"2023-01-30T22:02:30+00:00","remoteAddr":"","user":"[user redacted]","app":"workflow_ocr","method":"","url":"--","message":"OCR for file /[user redacted]/files/Misc-JD/OCR-Testing/iiiii - Copy - Copy - Copy.pdf not possible. Message: OCRmyPDF did not produce any output","userAgent":"--","version":"25.0.3.2","data":{"app":"workflow_ocr"}}
{"reqId":"zo00fPcm1UTyaxeOOfXe","level":0,"time":"2023-01-30T22:02:30+00:00","remoteAddr":"","user":"--","app":"workflow_ocr","method":"","url":"--","message":"ENDED -- Run OCA\\WorkflowOcr\\BackgroundJobs\\ProcessFileJob job. Argument: {\"filePath\":\"/[user redacted]/files/Misc-JD/OCR-Testing/iiiii - Copy - Copy - Copy.pdf\",\"settings\":\"{\\\"languages\\\":[\\\"eng\\\"],\\\"ocrMode\\\":1,\\\"removeBackground\\\":false}\"}.","userAgent":"--","version":"25.0.3.2","data":{"app":"workflow_ocr"}}
{"reqId":"zo00fPcm1UTyaxeOOfXe","level":0,"time":"2023-01-30T22:02:30+00:00","remoteAddr":"","user":"--","app":"cron","method":"","url":"--","message":"Finished OCA\\WorkflowOcr\\BackgroundJobs\\ProcessFileJob job with ID 12930 in 0 seconds","userAgent":"--","version":"25.0.3.2","data":{"app":"cron"}}
{"reqId":"zo00fPcm1UTyaxeOOfXe","level":0,"time":"2023-01-30T22:02:30+00:00","remoteAddr":"","user":"--","app":"cron","method":"","url":"--","message":"CLI cron call has selected job with ID 12931","userAgent":"--","version":"25.0.3.2","data":{"app":"cron"}}
{"reqId":"zo00fPcm1UTyaxeOOfXe","level":0,"time":"2023-01-30T22:02:30+00:00","remoteAddr":"","user":"--","app":"cron","method":"","url":"--","message":"Run OCA\\DAV\\BackgroundJob\\UploadCleanup job with ID 12931","userAgent":"--","version":"25.0.3.2","data":{"app":"cron"}}
{"reqId":"zo00fPcm1UTyaxeOOfXe","level":0,"time":"2023-01-30T22:02:30+00:00","remoteAddr":"","user":"--","app":"cron","method":"","url":"--","message":"Finished OCA\\DAV\\BackgroundJob\\UploadCleanup job with ID 12931 in 0 seconds","userAgent":"--","version":"25.0.3.2","data":{"app":"cron"}}
```
Any suggestions/help would be much appreciated!  Thx

**Comment by R0Wi** [comment]
Hi @farhills, I didn't test this but if you say that in genereal `ocrmypdf` works fine inside of your container, I guess you're facing some kind of permission issue. Depending on how Nextcloud is executed inside of your container, you should try to figure out under which user account the cron execution of Nextcloud is running. This depends on how you regularly execute the cron (via crontab, webcron, ...). Try to login with exactly that user into your container (something like `docker exec -it -u <USER> <CONTAINER> bash`) and run the `ocrmypdf` command again. I'd expect it to fail with the same error message? If that's the case you could either try to run your NC environment/cron as your regular user (where you did not face any issues) or try to fix your `ocrmypdf` installation, so that the NC user is able to use it.

Btw: do you use the official Nextcloud Docker Alpine container? And how did you install `ocmypdf` (via packages or manually)?

**Comment by farhills** [comment]
Success!  Thanks for the pointer re. permissions, I was logged in a root when it worked, and as user `abc` it failed with the same error shown in the nextcloud logs.  The container is linuxserver.

I am installing from packages via a bash script the executes on container creation:
```
#!/usr/bin/with-contenv bash

echo "**** OCR Dependencies SCRIPT START ****"
apk add --verbose --no-cache --upgrade \
        tesseract-ocr \
        ocrmypdf \
        python3 \
        py3-pip 

# pip install --user ocrmypdf

echo "**** CUSTOM SCRIPT END ****"
```

During my initial trials with the script I noticed that ocrmypdf was only at v12.x, not 14.  I added the `pip install` line in an effort to install from a different package manager and get it on the latest version (following some bad google advice), before I realized that 12.x is max for alpine 3.15 (the container).  Never deleted `pip install` from the script.  After spending far too much time blindly running `chmod` and `chown` inside the container, I discovered that `pip install` should not be run as root as it clobbers permissions and can break installations.  Bingo!

**Reply by farhills** [reply]
One other hitch that added confusion - the 'rename file' flow trigger appears to be broken.  It wasn't until I started watching the oc_jobs table that I realized OCR wasn't even being called.

**Reply by R0Wi** [reply]
Glad to hear that things are working now. Btw I'd recommend a ocrmypdf version pre 13 since they recently removed the `remove-background` flag.

Regarding the file rename: unfortunately the trigger section is a core part of every workflow app and belongs to the Nextcloud core, not the app itself. What I recommend here is to lower your loglevel and observe the server logs. The OCR app logs quite a few things, that might point you to the right direction.
